spring-cloud-starter-stream-common-cdc-debezium/spring-cloud-starter-stream-common-cdc-debezium-core/src/main/java/org/springframework/cloud/stream/app/cdc/common/core/CdcStreamConfiguration.java

```diff
@@ -56,8 +56,17 @@
 
 			Consumer<SourceRecord> messageConsumer = sourceRecord -> {
 
+				if (sourceRecord == null) {
+					logger.debug("Null sourceRecord");
+					return;
```

Comment:
Why do we just ignore?
Why the same `"null".getBytes()` or `byte[0]` doesn't work for us?

On the other thought: maybe we should emit some `ApplicationEvent` to say that source of events is exhausted and we are not going to receive any update afterwards.

Or do I miss anything?

Thanks

Reply:
Good question which is the core of this issue.  I will try to explain the context and my reasoning the implemented as above. 

Most or perhaps all of the  Debezium connectors have the concept of a `tombstone event` that can be controlled via the `cdc.flattering.dropTombstones` property (drop is true by default). 

The tombstone property controls whether a tombstone event should be generated after a row delete event. When enabled the delete operations are represented by a delete event and a subsequent tombstone event. if the tombstone is disabled only a delete event is sent only.

The tombstone event is particularly important in the context of Kafka. 
> Emitting the tombstone event allows Kafka to completely delete all events pertaining to the given key once the source record got  deleted. 

This raises the question if the tombstone concept is applicable for the EmbeddedDebezium use case we use. But since i don't know the answer i decided not to completely remove this concept although it is disabled by default. 

It turned out though that even if the tombstone is disabled the framework will send a second event to the messageConsumer handler but the value of the `sourceRecord` is null.  This is the reason i've added the ignore section in case of null sourceRecord.  So i had to explicitly filter out the second event. 

Next it turned out that when the tombstone event is enabled the serialization produces a `null` payload and all useful info is encoded into the `cdc.key` header.  But the MessageBuilder doesn't allow null payloads so i've put the "null" string. Which is bad value i guess. 
I should use an empty string instead?

Reply:
Well, thanks, @tzolov , for detailed explanation.

Right, `tombstone` in terms for Apache Kafka really means to remove the `key` from the log.
We use special `KafkaNull` constant and respective `KafkaNullAwarePayloadArgumentResolver` in Spring for Apache Kafka to deal with a `tombstone` concept.

Not sure if we need to introduce something similar here to have as much logic coverage as possible, but this some kind of food for you to think more.

An empty string is going to be treated as a `null payload` in some method argument resolver and therefore a logic might not be called again. But at the same time the `byte[0]` is treated same way:

```
	/**
	 * Specify if the given {@code payload} is empty.
	 * @param payload the payload to check (can be {@code null})
	 */
	protected boolean isEmptyPayload(@Nullable Object payload) {
		if (payload == null) {
			return true;
		}
		else if (payload instanceof byte[]) {
			return ((byte[]) payload).length == 0;
		}
		else if (payload instanceof String) {
			return !StringUtils.hasText((String) payload);
		}
		else {
			return false;
		}
	}
```

That's why I suggest to emit an `ApplicationEvent` instead of attemting to send a message with an *empty payload*.

Thanks